File: 9-Stack/programs/ruby/program2.rb
class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class Stack
  def initialize
    @top = nil
  end

  def push(x)
    t = Node.new(x)
    if t.nil?
      puts "Stack is full"
    else
      t.next = @top
      @top = t
    end
  end

  def pop
    x = -1
    if @top.nil?
      puts "Stack is Empty"
    else
      t = @top
      @top = @top.next
      x = t.data
    end
    x
  end

  def display
    p = @top
    while !p.nil?
      print "#{p.data} "
      p = p.next
    end
    puts
  end

  def peek(pos)
    p = @top
    (pos - 1).times do
      p = p.next unless p.nil?
    end
    p.nil? ? -1 : p.data
  end

  def stack_top
    @top.nil? ? -1 : @top.data
  end

  def is_empty
    @top.nil? ? 1 : 0
  end

  def is_full
    t = Node.new(0)
    r = t.nil? ? 1 : 0
    r
  end
end

stack = Stack.new
stack.push(10)
stack.push(20)
stack.push(30)
stack.display
puts stack.pop
stack.display
puts "Element at position 2 is: #{stack.peek(2)}"
puts "Top element is: #{stack.stack_top}"
puts "Is stack empty: #{stack.is_empty}"
puts "Is stack full: #{stack.is_full}"

File: 9-Stack/programs/cpp/program2.cpp
// Stack implementation

#include<bits/stdc++.h>
using namespace std;

struct Node{
  int data;
  struct Node *next;
}*top = NULL;

void push(int x){
  struct Node *t;
  t = new Node;
  if(t == NULL){
    cout << "Stack is full" << endl;
  }else{
    t->data = x;
    t->next = top;
    top = t;
  }
}

int pop(){
  int x = -1;
  if(top == NULL){
    cout << "Stack is Empty" << endl;
  }else{
    struct Node *t = top;
    top = top->next;
    x = t->data;
    delete t;
  }
  return x;
}

void Display(){
  struct Node *p = top;
  while(p != NULL){
    cout << p->data << " ";
    p = p->next;
  }
  cout << endl;
}

int peek(int pos){
  struct Node *p = top;
  for(int i = 0; p != NULL && i < pos-1; i++){
    p = p->next;
  }
  if(p != NULL){
    return p->data;
  }else{
    return -1;
  }
}


int stackTop(){
  if(top != NULL){
    return top->data;
  }else{
    return -1;
  }
}

int isEmpty(){
  return top ? 0 : 1;
}

int isFull(){
  struct Node *t = new Node;
  int r = t ? 0 : 1;
  delete t;
  return r;
}


int main(){
  push(10);
  push(20);
  push(30);
  Display();
  cout << pop() << endl;
  Display();
  cout << "Element at position 2 is: " << peek(2) << endl;
  cout << "Top element is: " << stackTop() << endl;
  cout << "Is stack empty: " << isEmpty() << endl;
  cout << "Is stack full: " << isFull() << endl;


  return 0;
}



